Composer/packages/server/src/models/bot/process/orchestratorWorker.ts

```diff
@@ -25,9 +26,89 @@ export class LabelResolversCache {
+    this.projects.clear();
+  }
+}
+
+export const cache = new LabelResolversCache();
+
+/**
+ * Orchestrator: Warm up the LabelResolversCache if .blu files already exist.
+ *
+ * The Orchestrator build process is iterative - the results of every build are cached, and the cache
+ * is used in subsequent builds to reduce the number of utterance embeddings that have to be re-calculated.
+ *
+ * However, if a user starts a new session of Composer and reopens the same bot project,
+ * the caches will be empty and training will begin from scratch again.
+ *
+ * If a user has ever built a bot with Orchestrator, embeddings (in the form of .blu files) for each
+ * utterance will be stored in the /generated folder.
+ *
+ * We warm up the LabelResolversCache with these blu files and pass this cache to the normal build
+ * process.  Re-hydrating the cache from files is still cheaper than recalculating the embeddings from scratch.
+ *
+ * @param projectId
+ * @param modelPath
+ * @param storage
+ * @param generatedFolderPath
+ */
+export async function warmUpCache(generatedFolderPath: string, projectId: string) {
+  //warm up the cache only if it's empty and we've built this bot before
+  if (!(await pathExists(generatedFolderPath)) || cache.get(projectId).size > 0) {
+    return false;
+  }
+
+  const bluFiles = (await readdir(generatedFolderPath)).filter((fileName) => fileName.endsWith('.blu'));
+
+  if (!bluFiles.length) {
+    return false;
+  }
+
+  const orchestratorSettingsPath = Path.resolve(generatedFolderPath, 'orchestrator.settings.json');
+  if (!(await pathExists(orchestratorSettingsPath))) {
+    return false;
+  }
+
+  // an implementation detail is that we need to use the right model to reproduce the right LabelResolvers
+  // so we get the model versions from a pre-existing settings file, and split the files based on
+  // language
+  const orchestratorSettings: IOrchestratorSettings = await readJson(orchestratorSettingsPath);
+  if (!orchestratorSettings?.orchestrator?.models || !orchestratorSettings?.orchestrator?.models) {
+    return false;
+  }
+
+  const [enLuFiles, multiLangLuFiles] = partition(bluFiles, (f) => f.split('.')?.[1].startsWith('en'));
+
+  const modelDatas = [
+    { model: orchestratorSettings?.orchestrator?.models?.en, lang: 'en', luFiles: enLuFiles },
+    { model: orchestratorSettings?.orchestrator?.models?.multilang, lang: 'multilang', luFiles: multiLangLuFiles },
+  ];
+
+  const [enMap, multilangMap] = await Promise.all(
+    modelDatas.map(async (modelData) => {
+      const snapshotData = await Promise.all(
+        modelData.luFiles.map(
+          async (f) =>
+            [f.replace('.blu', '.lu'), new Uint8Array(await readFile(Path.join(generatedFolderPath, f)))] as [
+              string,
+              Uint8Array
+            ]
+        )
+      );
+
+      return modelData.model && snapshotData.length
+        ? await Orchestrator.getLabelResolversAsync(modelData.model, '', new Map(snapshotData), false)
```

Comment:
I assume this matches the lu object id "f.replace('.blu', '.lu')" pass in the buildAsync method, correct?

Reply:
Yes, these are also the ids we get back from the results object from `buildAsync`,